pkg/model/awsmodel/api_loadbalancer.go

```diff
@@ -301,85 +302,118 @@ func (b *APILoadBalancerBuilder) Build(c *fi.ModelBuilderContext) error {
-			Lifecycle:     b.SecurityLifecycle,
-			CIDR:          fi.String("0.0.0.0/0"),
-			Egress:        fi.Bool(true),
-			SecurityGroup: lbSG,
-		}
-		AddDirectionalGroupRule(c, t)
-	}
-
-	// Allow traffic into the ELB from KubernetesAPIAccess CIDRs
-	if b.APILoadBalancerClass() == kops.LoadBalancerClassClassic {
-		for _, cidr := range b.Cluster.Spec.KubernetesAPIAccess {
+		{
 			t := &awstasks.SecurityGroupRule{
-				Name:          fi.String("https-api-elb-" + cidr),
+				Name:          fi.String("ipv4-api-elb-egress"),
 				Lifecycle:     b.SecurityLifecycle,
-				CIDR:          fi.String(cidr),
-				FromPort:      fi.Int64(443),
-				Protocol:      fi.String("tcp"),
+				CIDR:          fi.String("0.0.0.0/0"),
+				Egress:        fi.Bool(true),
 				SecurityGroup: lbSG,
-				ToPort:        fi.Int64(443),
 			}
 			AddDirectionalGroupRule(c, t)
-
-			// Allow ICMP traffic required for PMTU discovery
-			c.AddTask(&awstasks.SecurityGroupRule{
-				Name:          fi.String("icmp-pmtu-api-elb-" + cidr),
+		}
+		{
+			t := &awstasks.SecurityGroupRule{
+				Name:          fi.String("ipv6-api-elb-egress"),
 				Lifecycle:     b.SecurityLifecycle,
-				CIDR:          fi.String(cidr),
-				FromPort:      fi.Int64(3),
-				Protocol:      fi.String("icmp"),
+				IPv6CIDR:      fi.String("::/0"),
+				Egress:        fi.Bool(true),
 				SecurityGroup: lbSG,
-				ToPort:        fi.Int64(4),
-			})
+			}
+			AddDirectionalGroupRule(c, t)
 		}
 	}
 
-	masterGroups, err := b.GetSecurityGroups(kops.InstanceGroupRoleMaster)
-	if err != nil {
-		return err
-	}
-
-	if b.APILoadBalancerClass() == kops.LoadBalancerClassNetwork {
+	// Allow traffic into the ELB from KubernetesAPIAccess CIDRs
+	if b.APILoadBalancerClass() == kops.LoadBalancerClassClassic {
 		for _, cidr := range b.Cluster.Spec.KubernetesAPIAccess {
-
-			for _, masterGroup := range masterGroups {
+			{
 				t := &awstasks.SecurityGroupRule{
-					Name:          fi.String(fmt.Sprintf("https-api-elb-%s", cidr)),
+					Name:          fi.String("https-api-elb-" + cidr),
 					Lifecycle:     b.SecurityLifecycle,
-					CIDR:          fi.String(cidr),
 					FromPort:      fi.Int64(443),
 					Protocol:      fi.String("tcp"),
-					SecurityGroup: masterGroup.Task,
+					SecurityGroup: lbSG,
 					ToPort:        fi.Int64(443),
 				}
+				if utils.IsIPv6CIDR(cidr) {
+					t.IPv6CIDR = fi.String(cidr)
+				} else {
+					t.CIDR = fi.String(cidr)
+				}
 				AddDirectionalGroupRule(c, t)
+			}
 
-				// Allow ICMP traffic required for PMTU discovery
+			// Allow ICMP traffic required for PMTU discovery
+			if utils.IsIPv4CIDR(cidr) {
 				c.AddTask(&awstasks.SecurityGroupRule{
 					Name:          fi.String("icmp-pmtu-api-elb-" + cidr),
 					Lifecycle:     b.SecurityLifecycle,
 					CIDR:          fi.String(cidr),
 					FromPort:      fi.Int64(3),
 					Protocol:      fi.String("icmp"),
-					SecurityGroup: masterGroup.Task,
+					SecurityGroup: lbSG,
 					ToPort:        fi.Int64(4),
 				})
+			}
```

Comment:
Don't we need to open type 2 ICMPv6 traffic for IPv6 PMTU discovery?

Reply:
...or even open all of ICMPv6?

Reply:
I tried doing this manually and got into some trouble with the AWS web UI, which seems to not allow setting custom ICMPv6, but seems to work from API.